### PR TITLE
Ux

### PR DESCRIPTION
**Pull request by idanstark42**
So... what did I do here?
1. Restructured the HTML pages, and added adequate, responsive CSS instead of the old UI. Now there are 4 pages: 
 - Scoresheet
 - Scores - which includes the rankings as well
 - Tournament - which includes all tournament settings
 - Settings, which include more advanced options
Removed all confusion options, and split the scoresheet into 4 steps:
 - intro - only once, setting the ref and table, 
 - team selection
 - missions form
 - checkout
After the checkout the user is referenced back to teams selection. The "edit" button in the scores page references the user to the "missions form" step here.
2. Created and used across the system the `ng-datatable` directive, which displays data in an easy to review, sortable, searchable way, with infinite scrolling for heavier data.
3. Split the CSS into generic CSS, used in every season, and CSS per challenge, giving it a theme.
4. Added the "stay on the same page" functionality: When you enter a page, it's saved in the `path` part of the URL, so that if you refresh it will go to the same page.
5. Performance improvments: removed various instances of `$watch`, changed `ng-show` to `ng-if` everywhere that it should be (pages for instance), Added the `dialogs` service, which makes the usage of handshakes redundent.

That's it all in all, I hope you like it. There's a lot left to improve here.

### CODE REVIEW

File: src/js/directives/datatable.js [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+                scope.onCellClick = function(item, column) {
+                    if(column.onCellClick) {
+                        column.onCellClick(item);
+                    } else if(scope.edit && !scope.edit.is(item, column)) {
+                        dontLoad = true;
+                        scope.edit.start(item, column);
+                    }
+                };
+
+                scope.scroll = (function() {
```
[the remark this answers is not in the document]
Comment (ThinkRedstone) [reply src/js/directives/datatable.js]: Also, it seems to render the loading animation only after in finishes loading on mobile, resulting in the table freezing while it's loading and then displaying the loading animation for a split second.

File: src/views/dialogs.html [hunk 2]
```diff
@@ -1,198 +1,51 @@
...
-                    <label for="">Rounds (0 to disable stage)</label>
-                </p>
-                <p>
-                    <button class="btn" ng-click="ok()" ng-disabled="!stageDefValid()">Ok</button>
-                    <button class="btn" ng-click="cancel()">Cancel</button>
-                </p>
-            </div>
-        </section>
+        <div class="dialog-actions">
+            <button class="btn btn-primary" ng-click="ok()">Import</button>
```
Comment (ThinkRedstone) [inline src/views/dialogs.html]: you renamed `ok()` to `save()` so this `ng-click` matches no function

File: src/views/pages/clock.html [hunk 3]
```diff
@@ -1,32 +0,0 @@
-<div ng-controller="clockCtrl">
```
Comment (ThinkRedstone) [inline src/views/pages/clock.html]: again, this is needed by the scorekeeper to start the clock

File: src/views/appbar.html [hunk 4]
```diff
@@ -0,0 +1,7 @@
+<div class="appbar">
+    <i id="menu-icon" ng-click="toggleDrawer()" class="big material-icons">menu</i>
+    <h3>
+        <i>FIRST</i> LEGO Legaue Scoring system
+    </h3>
+    <a class="btn logout" ng-if="user" href="/logout" ><i class="material-icons">exit_to_app</i> <span>Logout</span></a>
```
Comment (ThinkRedstone) [inline src/views/appbar.html]: this is a problematic place to put the logout button, because if you only have access to one page you don't see the appbar, and so can't logout (refs can't logout, for example)

File: src/views/pages/scoresheet/intro.html [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    <div class="fluid">
+    	<select ng-model="scoreEntry.table" ng-options="item.name as item.name for item in settings.tables"></select>
+    </div>
+
+    <label>Referee:</label>
+    <div class="fluid">
+    	<select ng-model="scoreEntry.referee" ng-options="item.name as item.name for item in settings.referees"></select>
+    </div>
+
+	<div class="center">
```
Comment (ThinkRedstone) [inline src/views/pages/scoresheet/intro.html]: mismatched indentation: mixed usage of tabs and spaces

File: src/views/pages/settings.html [hunk 6]
```diff
@@ -1,236 +1,58 @@
...
-                            Broadcast 0 for negative scores
-                        </label>
-                    </p>
+                    <div class="setting">
+                        <div class="setting-title">Challenge Host</div>
+                        <input type="text" ng-model="settings.host" ng-change="save()">
+                    </div>
+
+                <div class="setting">
+                    <div class="setting-title">Mhub Host</div>
```
Comment (ThinkRedstone) [inline src/views/pages/settings.html]: You're missing here the delimiter settings for ranking exports, and "enable custom mhub" setting

File: src/views/pages/tournament.html [hunk 7]
```diff
@@ -0,0 +1,44 @@
+<div ng-controller="tournamentCtrl" id="tournament">
+
+    <div class="teams card">
+        <div class="card-top">
+            <div class="card-header" ng-click="show.teams = !show.teams">
+                <i class="material-icons">{{showIcon(teams.show)}}</i>
```
Comment (ThinkRedstone) [inline src/views/pages/tournament.html]: this should be `show.teams` not `teams.show`

File: src/views/pages/tournament.html [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+                </div>
+                <div class="btn" ng-click="importTeams()">Import</div>
+            </div>
+        </div>
+        <ng-datatable id="teams-table" collection="teams" config="teamsTableConfig" ng-show="show.teams" class="scrollable-body"></ng-datatable>
+    </div>
+
+    <div class="structure card">
+        <div class="card-header" ng-click="show.structure = !show.structure">
+            <i class="material-icons">{{showIcon(structure.show)}}</i>
```
Comment (ThinkRedstone) [inline src/views/pages/tournament.html]: this should be `show.structure` not `structure.show`

File: src/views/pages/tournament.html [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+        <div class="card-body" ng-show="show.structure">
+            <ng-datatable id="stages-table" collection="stages" config="stagesTableConfig" class="scrollable-body"></ng-datatable>
+            <ng-datatable id="referees-table" collection="settings.referees" config="refereesTableConfig" class="scrollable-body"></ng-datatable>
+            <ng-datatable id="tables-table" collection="settings.tables" config="tablesTableConfig" class="scrollable-body"></ng-datatable>
+        </div>
+    </div>
+
+    <div class="challenge card">
+        <div class="card-header" ng-click="show.challenge = !show.challenge">
+            <i class="material-icons">{{showIcon(challenge.show)}}</i>
```
Comment (ThinkRedstone) [inline src/views/pages/tournament.html]: this should be `show.challenge` not `challenge.show`

File: src/css/tournament.css [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+}
+
+#tounament ng-datatable td.create-message {
+    font-size: 1.5rem !important;
+    text-align: center;
+    cursor: pointer;
+}
+
+#teams-table tbody {
+    height: 30vh;
```
Comment (ThinkRedstone) [inline src/css/tournament.css]: You should make this taller, so more teams can be viewed at once; right now only less than half the height of the screen is being used to view teams (now that I think about it, only exactly 30 percent)

File: src/views/dialogs.html [hunk 11]
```diff
@@ -1,198 +1,51 @@
...
-                </tr>
-            </table>
-
-
-            <div>
-                <button class="btn" ng-click="ok()">Ok</button>
-                <button class="btn" ng-click="cancel()">Cancel</button>
+             <h4>Teams</h4>
+            <div id="teams-data" ng-if="importLines">
+                <table>
```
Comment (ThinkRedstone) [inline src/views/dialogs.html]: missing the `table` class, making it hard to read

File: src/css/dialog.css [hunk 12]
```diff
@@ -1,83 +1,51 @@
+
 .fll-dialog-shim {
```
[the remark this answers is not in the document]
Comment (idanstark42) [reply src/css/dialog.css]: Not mine, but sure

File: src/components/angular-animate/angular-animate.js [hunk 13]
```diff
@@ -0,0 +1,1689 @@
+/**
+ * @license AngularJS v1.2.20
+ * (c) 2010-2014 Google, Inc. http://angularjs.org
+ * License: MIT
```
[the remark this answers is not in the document]
Comment (idanstark42) [reply src/components/angular-animate/angular-animate.js]: The `.gitignore` has an entry so that it wouldn't ignore this file

File: src/css/dialog.css [hunk 14]
```diff
@@ -1,83 +1,51 @@
...
-    margin-left: -300px;
-    margin-top: -200px;
-    border: 1px solid #999;
-    border: 1px solid rgba(0,0,0,0.3);
-    background-color: white;
-    border-radius: 6px;
-    box-shadow: 0 16px 48px rgba(0,0,0,0.6);
-    overflow: hidden;
+.fll-dialog-shim.ng-hide {
+    display: block !important;
```
[the remark this answers is not in the document]
Comment (idanstark42) [reply src/css/dialog.css]: Yes. `!important` is considred bad practice because it overrides other CSS. It applies to the situation in whcih you write the entire CSS, but that's not our case. The `!improtant` clause is used here to override angular or bootstrap CSS to customize it to the program's needs.

File: src/css/dialog.css [hunk 15]
```diff
@@ -1,83 +1,51 @@
...
-
-/* display back arrow on fullscreen dialogs */
-.smallWindow .fll-dialog-back {
-    display: inline-block;
-}
-.smallWindow .fll-dialog-close {
-    display: none;
+.dialog-actions {
+    margin-top: .75em;
+    direction: rtl;
```
[the remark this answers is not in the document]
Comment (idanstark42) [reply src/css/dialog.css]: Good point. I'll see if flexbox works better here

File: src/css/drawer.css [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+  left: 0;
+  width: 10em;
+  height: calc(100vh - 3em);
+  z-index: 1;
+
+  background: grey;
+}
+
+.drawer.ng-hide {
+  left: -10em;
```
[the remark this answers is not in the document]
Comment (idanstark42) [reply src/css/drawer.css]: It's for animation effects

File: src/css/drawer.css [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+.drawer a {
+  cursor: pointer;
+	color: inherit;
+}
+
+.drawer a:hover {
+    background-color: black !important;
```
[the remark this answers is not in the document]
Comment (idanstark42) [reply src/css/drawer.css]: Same reason

File: src/css/elements.css [hunk 18]
```diff
@@ -1,345 +1,235 @@
...
+    0% { transform: rotate(0deg);  }
+    100% { transform: rotate(360deg);  }
+}
+
+@keyframes spin-counter-clockwise {
+    0% { transform: rotate(0deg);  }
+    100% { transform: rotate(-360deg);  }
+}
+
+.dimmer {
```
[the remark this answers is not in the document]
Comment (idanstark42) [reply src/css/elements.css]: Sure. will do

File: src/css/elements.css [hunk 19]
```diff
@@ -1,345 +1,235 @@
...
+    margin: 0;
+}
+
+.card .card-top {
+    display: flex;
+    flex-direction: row;
+    justify-content: space-between;
+}
+
+.card .card-header i {
```
[the remark this answers is not in the document]
Comment (idanstark42) [reply src/css/elements.css]: Good point. Will fix this

File: src/css/elements.css [hunk 20]
```diff
@@ -1,345 +1,235 @@
...
+    margin: 0;
+}
+
+.card .card-body {
+    padding: 1em;
+}
+
+/* ng-datatble directive global styles */
+
+ng-datatable table {
```
[the remark this answers is not in the document]
Comment (idanstark42) [reply src/css/elements.css]: I didn't understand the first half, bu the second half is easily fixable with `cursor: pointer`

File: src/css/pages.css [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+
+/* appbar */
+
+.appbar {
+    display: flex;
+    align-items: center;
+    justify-content: space-between;
+	height: 3em;
```
[the remark this answers is not in the document]
Comment (idanstark42) [reply src/css/pages.css]: Will fix this

File: src/css/pages.css [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+
+}
+
+@media screen and (min-width: 100vh) {
+	#menu-icon {
+		display: none;
+	}
+
+	.page:not(.only-page) {
+		margin-left: 10em;
```
[the remark this answers is not in the document]
Comment (idanstark42) [reply src/css/pages.css]: It's for animations

File: src/css/scoresheet.css [hunk 23]
```diff
@@ -1,148 +1,277 @@
...
+#scoresheet-teams.ng-hide {
+    opacity: 0;
+    transition: none;
 }
 
-.missions .field {
+#scoresheet-teams .stage-select .round-score {
+    font-weight: bolder;
+    font-size: 1.5rem;
+    transform: rotate(-20deg);
```
[the remark this answers is not in the document]
Comment (idanstark42) [reply src/css/scoresheet.css]: Maybe. That was a style choice. I'll ask the UX proffessional what she thinks

File: src/css/scoresheet.css [hunk 24]
```diff
@@ -1,148 +1,277 @@
...
+    transition: none;
+}
+
+/* Missions */
+
+#missions {
+    overflow-y: auto;
+}
+
+#missions .mission .objective {
```
[the remark this answers is not in the document]
Comment (idanstark42) [reply src/css/scoresheet.css]: Sort of... the UX wasn't built with screens this small in mind

File: src/css/scoresheet.css [hunk 25]
```diff
@@ -1,148 +1,277 @@
...
+
+#signoff .signoff-grid .missions .objective .title {
+    display: inline-block;
+    font-weight: bold;
 }
-.scorereview {
-	padding: 10px 2px 10px 10px;
-	margin: 2px;
+
+#signoff .signoff-grid .missions .objective .value {
```
[the remark this answers is not in the document]
Comment (idanstark42) [reply src/css/scoresheet.css]: Sure

File: src/css/settings.css [hunk 26]
```diff
@@ -0,0 +1,64 @@
+#settings {
+    height: inherit;
+    display: flex;
+    flex-direction: column;
+}
+
+#settings > *:first-child {
```
[the remark this answers is not in the document]
Comment (idanstark42) [reply src/css/settings.css]: It selects the first child of the element with id `#settings`. The idea of a selector is to follow the logic of the selected element. I want this to work on whatever element goes in there first. In this case, it means that `#settings` page can have only one child, since it taked the entire height of it's parent

File: src/css/settings.css [hunk 27]
```diff
@@ -0,0 +1,64 @@
...
+    height: inherit;
+    display: flex;
+    flex-direction: column;
+}
+
+#settings .tab-content {
+	flex: 1;
```
[the remark this answers is not in the document]
Comment (idanstark42) [reply src/css/settings.css]: Will fix that

File: src/css/settings.css [hunk 28]
```diff
@@ -0,0 +1,64 @@
...
+    display: flex;
+    flex-direction: column;
+}
+
+#settings .tab-content .tab-pane.active {
+	flex: 1;
```
[the remark this answers is not in the document]
Comment (idanstark42) [reply src/css/settings.css]: Will fix that

File: src/css/themes/2018_hydrodynamics.css [hunk 29]
```diff
@@ -0,0 +1,233 @@
...
+    border-radius: 2px;
+}
+
+ng-datatable thead {
+    width: calc(100% - 10px) !important;
+}
+
+.appbar, .drawer {
+	color: #EDFAFD; /* Glass */
+	background: #135589; /* Marina */
```
[the remark this answers is not in the document]
Comment (idanstark42) [reply src/css/themes/2018_hydrodynamics.css]: Will fix that

File: src/css/themes/2018_hydrodynamics.css [hunk 30]
```diff
@@ -0,0 +1,233 @@
...
+#missions .mission.error, #missions .errors.card {
+	background: #FFB1B1;
+}
+
+#missions .mission.error .errors {
+    border: #8c3030 1px solid;
+    padding: 0.2em;
+}
+
+#missions .mission .mission-score {
```
[the remark this answers is not in the document]
Comment (idanstark42) [reply src/css/themes/2018_hydrodynamics.css]: No, it was intended. Only styles that are relevant only to this year should be in here

File: src/css/tournament.css (hunk 10 above)
Comment (idanstark42) [reply src/css/tournament.css]: Yes, this was intendent. I'll ask the UX prof what she thinks

File: src/js/controllers/TeamImportDialogController.js [hunk 31]
```diff
@@ -71,48 +70,38 @@ define('controllers/TeamImportDialogController', [
                         $scope.delimiter = ",";
                         $scope.importRaw = data;
                     }
+                    parseData();
```
[the remark this answers is not in the document]
Comment (idanstark42) [reply src/js/controllers/TeamImportDialogController.js]: Will fix that

File: src/js/controllers/TeamImportDialogController.js [hunk 32]
```diff
@@ -71,48 +70,38 @@ define('controllers/TeamImportDialogController', [
...
-
-            $handshake.$on('importTeams', function (e) {
-                $scope.dialogVisible = true;
-                defer = $handshake.defer();
-                return defer.promise;
-            });
-
-            $scope.ok = function () {
-                $scope.dialogVisible = false;
+            $scope.save = function () {
```
[the remark this answers is not in the document]
Comment (idanstark42) [reply src/js/controllers/TeamImportDialogController.js]: Will fix that

File: src/js/controllers/TeamImportDialogController.js [hunk 33]
```diff
@@ -71,48 +70,38 @@ define('controllers/TeamImportDialogController', [
...
-                defer.resolve({teams: teams});
+
+                if (teams) {
+                    $teams.clear();
+                    teams.forEach(function(team) {
+                        $teams.add({
+                            number: team.number,
+                            name: team.name
+                        });
+                    });
```
[the remark this answers is not in the document]
Comment (idanstark42) [reply src/js/controllers/TeamImportDialogController.js]: Will fix that

File: src/js/directives/datatable.html [hunk 34]
```diff
@@ -0,0 +1,96 @@
...
+                    <div class="btn" class="save" ng-click="create.reset()">
+                        <i class="material-icons">undo</i>
+                    </div>
+                    <div class="btn" class="save" ng-click="create.save()">
+                        <i class="material-icons">save</i>
+                    </div>
+                </div>
+            </td>
+
+            <td class="create-message" ng-show="create.showMessage" ng-click="create.disableMessage()">
```
[the remark this answers is not in the document]
Comment (idanstark42) [reply src/js/directives/datatable.html]: Good point. Will fix that

File: src/js/directives/datatable.js [hunk 35]
```diff
@@ -0,0 +1,209 @@
...
+                        element.addClass('dimmed');
+                        dontLoad = false;
+                    }
+                    $timeout(() => element.removeClass('dimmed'), 0, false);
+                    var collection = (scope.$parent.$eval(attrs.collection) || []);
+                    scope.collectionLength = collection.length;
+                    return collection;
+                };
+
+                scope.sort = {
```
[the remark this answers is not in the document]
Comment (idanstark42) [reply src/js/directives/datatable.js]: I dont understand... can you explian?

File: src/js/directives/datatable.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (idanstark42) [reply src/js/directives/datatable.js]: The mobile problem is interesting... Regarding the large collections problem: it must be such, since otherwise the large collection will be to heavy for the screen. This is the reason I introduced this the begin with

File: src/js/views/clock.js [hunk 36]
```diff
@@ -1,107 +0,0 @@
-define('views/clock', [
-    'services/log',
-    'services/ng-message',
-    'services/ng-settings',
-    'angular'
-], function (log) {
```
[the remark this answers is not in the document]
Comment (idanstark42) [reply src/js/views/clock.js]: It was temporary. After the fix you did in the clock's CSS, he can hold a smaller clock window on teh side of the screen and click it when he wishes

File: src/js/views/scores.js [hunk 37]
```diff
@@ -1,45 +1,125 @@
...
+                    }, {
+                        onClick: (score) => {
+                            togglePublished(score);
+                        },
+                        show: (score) => !score.published,
+                        icon: 'add_circle_outline'
+                    }, {
+                        onClick: (score) => {
+                            $scope.setPage($scope.pages.find(function (p) {return p.name === "scoresheet"}));
+                            $scope.$parent.goTo('scoresheet', $scoresheet => {
```
[the remark this answers is not in the document]
Comment (idanstark42) [reply src/js/views/scores.js]: That's a good point...

File: src/js/views/scoresheet.js [hunk 38]
```diff
@@ -212,23 +262,28 @@ define('views/scoresheet',[
...
+                $scope.scoreEntry.signature = null;
                 $scope.missions.forEach(function(mission) {
                     mission.objectives.forEach(function(objective) {
                         delete objective["value"];
                     });
                 });
 
                 if($scope.editingScore){
-                    $scope.setPage($scope.pages.find(function (p) {return p.name === "scores"}));
+                    $scope.goTo('scores');
```
[the remark this answers is not in the document]
Comment (idanstark42) [reply src/js/views/scoresheet.js]: Will fix that

File: src/js/views/scoresheet.js [hunk 39]
```diff
@@ -212,23 +262,28 @@ define('views/scoresheet',[
...
             };
 
-            $scope.saveEdit = function () {
-                $scope.setPage($scope.pages.find(function (p) {return p.name === "scores"}));//When you finish editing a scoresheet, it returns you to the scores view
+            $scope.save = function() {
+                $scope.editingScore ? update() : create();
+            }
+
+            function update() {
+                $scope.goTo('scores');//When you finish editing a scoresheet, it returns you to the scores view
```
[the remark this answers is not in the document]
Comment (idanstark42) [reply src/js/views/scoresheet.js]: Will fix that

File: src/js/views/scoresheet.js [hunk 40]
```diff
@@ -280,59 +334,42 @@ define('views/scoresheet',[
...
-                });
-            };
-
             $scope.loadScoresheet = function (score) {
                 log(`Editing scoresheet: stage ${score.stageId}, round ${score.round}, team ${score.teamNumber}, score ${score.score}`);
                 $scope.editingScore = true;
                 $scope.scoreEntry = score;
                 $scores.loadScoresheet(score).then(function (result) {
-                    $scope.signature = result.signature;
-                    $scope.referee = result.referee;
```
[the remark this answers is not in the document]
Comment (idanstark42) [reply src/js/views/scoresheet.js]: Will fix that

File: src/js/views/tournament.js [hunk 41]
```diff
@@ -0,0 +1,220 @@
...
+                    save: (newTeam) => {
+                        $teams.add(newTeam);
+                        $teams.save();
+                    }
+                },
+                row: {
+                    classes: (team) => `team_${team.number}`
+                },
+                search : () => $scope.teamsSearch,
+                scrollCount: 10
```
[the remark this answers is not in the document]
Comment (idanstark42) [reply src/js/views/tournament.js]: I can maybe up it a little, but it really gets stuck... There are a lot of performence issues to fix here

File: src/css/elements.css (hunk 20 above)
[the remark this answers is not in the document]
Comment (ThinkRedstone) [reply src/css/elements.css]: If you look at the picture I attached you can see I just added a team. However, aside from the undo and save buttons on the right, there is nothing indicating that I added a team, and it isn't obvious where I input the team number and and name (same goes for other places where you manually add entries to an ng-datatable, like stages). This happens because when I add a team, all fields are blank- so nothing renders aside from the buttons on the right.

File: src/js/directives/datatable.js (hunk 35 above)
[the remark this answers is not in the document]
Comment (ThinkRedstone) [reply src/js/directives/datatable.js]: because it uses the word sort in so many different ways (it has it's own property called `sort`, uses `attrConfig.sort`, `scope.sort.sort`) it's hard to follow. So renaming some properties like `scope.sort.sort` to `scope.sort.sortedColumn` or adding comments would be much appreciated.

File: src/js/directives/datatable.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (ThinkRedstone) [reply src/js/directives/datatable.js]: I'm not sure what "too heavy for the screen" means. Can't you load the large collection in the background? Display a small fraction as soon as it is ready, and load the rest in the meanwhile?

File: src/js/views/scores.js (hunk 37 above)
[the remark this answers is not in the document]
Comment (ThinkRedstone) [reply src/js/views/scores.js]: Just note this works because the fact that `onload` for that page can only happen after this function returns and frees the thread, so this still works even though though it shouldn't logically.

File: src/js/views/tournament.js [hunk 42]
```diff
@@ -0,0 +1,220 @@
...
+            };
+
+            $teams.init().then(function() {
+                $scope.teams = $teams._rawTeams;
```
[the remark this answers is not in the document]
Comment (idanstark42) [reply src/js/views/tournament.js]: Because of this line:
`return $http.post("/teams/save", { teams: this._rawTeams }).then(function (data, status) {`
in `Teams.prototype.save`. There's no regards for the `$teams.teams` collection. Splitting them in the first place creates negative performance. This is to save this trouble, and aspeccially for performance

File: src/views/appbar.html (hunk 4 above)
Comment (idanstark42) [reply src/views/appbar.html]: Yes, I knew that... I expect this will come from Yossi's review, perhaps with a new idea on where to put it, because I didn't have any...

File: src/views/dialogs.html (hunk 11 above)
Comment (idanstark42) [reply src/views/dialogs.html]: Will fix that

File: src/views/dialogs.html (hunk 2 above)
Comment (idanstark42) [reply src/views/dialogs.html]: Will fix that

File: src/views/pages/clock.html (hunk 3 above)
Comment (idanstark42) [reply src/views/pages/clock.html]: Same answer

File: src/views/pages/scoresheet/intro.html (hunk 5 above)
Comment (idanstark42) [reply src/views/pages/scoresheet/intro.html]: Will fix that

File: src/views/pages/tournament.html (hunk 7 above)
Comment (idanstark42) [reply src/views/pages/tournament.html]: Will fix that

File: src/views/pages/settings.html (hunk 6 above)
Comment (idanstark42) [reply src/views/pages/settings.html]: Yes, I had a lot of missing settings... Will fix that

File: src/views/pages/tournament.html (hunk 8 above)
Comment (idanstark42) [reply src/views/pages/tournament.html]: Will fix that

File: src/views/pages/tournament.html (hunk 9 above)
Comment (idanstark42) [reply src/views/pages/tournament.html]: Will fix that

File: src/css/elements.css (hunk 20 above)
[the remark this answers is not in the document]
Comment (idanstark42) [reply src/css/elements.css]: Ok, I get it. Doesn't it focus on the first field? It should...

[the remark this answers is not in the document]
Comment (idanstark42) [reply src/css/elements.css]: Will fix that

File: src/js/views/tournament.js (hunk 42 above)
[the remark this answers is not in the document]
Comment (ThinkRedstone) [reply src/js/views/tournament.js]: Then maybe change `$teams` to make it more efficient and fix this?

[the remark this answers is not in the document]
Comment (idanstark42) [reply src/js/views/tournament.js]: Yes, I didn't want to mess with the service's structure... This is more for a cleaning PR, which I guess will come next

File: src/js/directives/datatable.js (hunk 35 above)
[the remark this answers is not in the document]
Comment (idanstark42) [reply src/js/directives/datatable.js]: Will fix that. I will change the name of `sort.sort` to `sort.column`

File: src/js/directives/datatable.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (idanstark42) [reply src/js/directives/datatable.js]: `ng-reapet` in angular 1 is very heavy, since it constantly updating it. The usage of filters here is not helping but I didn't really have a choice under the scope of this technology. When I did it, I saw the performance problem that came with it, so I started looking for a solution, and the most popular one that most people reccommended was pagination or infinte scrolling. I thought infinte scrolling was more elegent

File: src/js/views/scores.js (hunk 37 above)
[the remark this answers is not in the document]
Comment (idanstark42) [reply src/js/views/scores.js]: I was really wonderring...

File: src/views/pages/settings.html (hunk 6 above)
Comment (idanstark42) [reply src/views/pages/settings.html]: This requires some thought as for the location of those settings and the correct seperation to tabs

**Comment by idanstark42** [comment]
I fixed most of the stuff you wrote. The things I didn't do yet:
 - Add missing settings (after thinking about their division into tabs)
 - Passing test
 Is there anything else you noticed?

File: src/js/views/scoresheet.js [hunk 43]
```diff
@@ -367,6 +367,7 @@ define('views/scoresheet',[
             $scope.loadScoresheet = function (score) {
                 log(`Editing scoresheet: stage ${score.stageId}, round ${score.round}, team ${score.teamNumber}, score ${score.score}`);
                 $scope.editingScore = true;
+                score.referee = $scope.scoreEntry.referee;
```
Comment (ThinkRedstone) [inline src/js/views/scoresheet.js]: Shouldn't this be the other way around? This lines override the referee of the loaded score with some other referee (in this case, the referee of the last score) but we want the referee of a loaded score to not change when editing it, unless it is specifically edited

Comment (idanstark42) [reply src/js/views/scoresheet.js]: The referee and the table here become a part of this client's environment until the next refresh. That means that every new score and every editing it will do will be his. This is important because the referee that edits need to have the responsibility for the edited version. I will add a fallback in case there's no referees set

Comment (ThinkRedstone) [reply src/js/views/scoresheet.js]: Again, before this change editing a score didn't change the referee unless you specifically changed it- i.e. if you changed the round for a specific score, the referee would stay the same. As far as I understand, this change changes this behaviour, which I feel is something we want to keep: it allows a scorekeeper to change a miss-typed mission and still keep the records of the referee that originally filled the scoresheet.

Comment (idanstark42) [reply src/js/views/scoresheet.js]: Yes, I know. I changed this functionality. The original idea of the referee saved in the scoresheet is to have someone to ask about this scoresheet. Until we fully create an "editing history" feature, this should be the last person to change it, which is the current referee (only if they save it, of course). If we really disagree on this, I say we ask Alan and Yossi, since they have a better insight on functionality.

Comment (ThinkRedstone) [reply src/js/views/scoresheet.js]: I don't really care that much, It's just that it seems odd to me
